### PR TITLE
Array(T) implementation

### PR DESCRIPTION
**Pull request by waralex**
Array(T) implementation

### CODE REVIEW

File: src/columns/Array.jl
Comment (athre0z) [inline src/columns/Array.jl:84]: I think there `where {T}` is redundant here in modern Julia versions.
```suggestion
const PossibleVectors{T} =
     Union{<:AbstractVector{T}, <:AbstractCategoricalVector{T}}
```

**Review by athre0z — APPROVED** [review]
LGTM!

I annotated a few minor style related things, however I want to stress that these are more hints/tips on how I'd do things rather than things that need fixing / block the PR. I feel like I forced you into breaking lines and that you're not really used to it and thought they might be helpful, but I don't want to be the kind of maintainer who blocks PRs because some stuff is indented a bit weird. The same goes for any such style related comments that I might make in the future. Feel free to just leave these things as-is and I'll probably at some point do a sweep through the whole repo and clean it up (and drop it as a PR so you can flame me when you hate it! :) ).

File: src/columns/Array.jl
Comment (waralex) [reply src/columns/Array.jl]: fixed as 01b980e

**Comment by waralex** [comment]
>I feel like I forced you into breaking lines and that you're not really used to 

Be a little. But on the other hand I have slightly reduced the font size and now I have 3x2 editor windows instead of 2x2 :)

**Comment by waralex** [comment]
>Feel free to just merge these things as-is and I'll probably

Did I understand correctly that getting the `approval` status from you means that I can merge despite the remaining conversation?

**Comment by athre0z** [comment]
> I have slightly reduced the font size and now I have 3x2 editor windows instead of 2x2 :)

Heh, nice. Personally, I usually have two editors + one term but rather big font so that I can sit back nice and comfy at a bit of a distance from the screen.

> Did I understand correctly that getting the approval status from you means that I can merge despite the remaining conversation?

Yes. "approved" + remaining comments means that I consider the remaining stuff minor enough that it is very unlikely that this will need another review after you fixed them (e.g. fix a comment) or that they are merely hints that don't block the PR at all (e.g. style stuff).

On the matter of merges, I'd like to propose that we generally merge PRs via squash to get rid of all the minor fix-up commits, commits that get created when you accept suggestions in the GH GUI, intermediate merges etc.. It is also possible to do interactive rebases (via a desktop client / CLI) and then do a regular merge, where you can clean up all this stuff on a commit by commit level. However, if you're not already familiar with that workflow, I will freely admit that it has quite a steep learning curve and there is a high chance to mess things up until you're used to it, so when in doubt, just go with the squash. :)

**Comment by waralex** [comment]
>However, if you're not already familiar with that workflow, I will freely admit that it has quite a steep learning curve and there is a high chance to mess things up until you're used to it

I've never done this before, but I can't get past an issue that "has a pretty steep learning curve" and not try it:). Can you give me a link to some guide that you can start with?

**Comment by athre0z** [comment]
It depends a bit on what client you're using. For a guide using the CLI, this one looks quite solid:
https://matthewsetter.com/git-interactive-rebase/

Also, the official doc:
https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

Another neat feature closely related:
https://blog.sebastian-daschner.com/entries/git-commit-fixup-autosquash

What makes the learning curve steep is mostly the fact that you need to force-push your changes after an interactive rebase (actually, any kind of rebase where you're rewriting history that you have previously pushed already) and you can break / change history on the remote branch. For this reason, before doing IBs for clean-up before merging, I will generally create a new branch from the branch that I intend to clean-up, do the IB on the copy and if something goes horribly wrong, I can just undo all changes by deleting the entire branch and creating a new one (instead of having to painfully fix it via reflog) and, when I'm happy, merge this into master. 

You can force-push onto feature branches for clean-up or rebase, but whatever you do, **never** force-push onto master -- it will break everyone's local copies! :p

Personally I use [TowerGit](https://www.git-tower.com/) these days which makes IBs a **lot** easier because you can mostly just drag & drop commits around as well as just deleting and rewriting any commit without any hassle, but it's paid software.

**Pull request merged**
